### DISCUSSION TITLE
Missing reference for keyboard keys

### DISCUSSION THREAD

**Discussion by ed2050** [zauberzeug/nicegui#2974] [Q&A] [answered]
### Question

Where's the reference for keyboard keys?  There should be a complete list of possible values for KeyEventArguments passed to ui.keyboard callback.  These are not provided anywhere in the [ui.keyboard doc page](https://nicegui.io/documentation/keyboard).

For instance:
- e.modifiers.shift is shown for shift key.  What other modifiers are available?  Different systems use different names: meta, option, window, etc.  A complete list is needed.
- arrow keys are shown as e.key.arrow_left, etc.  What other properties does e.key have for keystrokes?
- how do you detect unprintable keys like backspace, delete, home, end, etc?  Probably e.key.something... but what?

If these values are taken from another reference (such as js key events), then the complete spec should be linked from the ui.keyboard docs page.  Did I miss them somewhere else?

Thanks.

**Comment by falkoschindler** [comment, marked as answer]
Thanks for bringing this up, @ed2050. I just created PR #3010 to address this issue.

**Reply by ed2050** [reply]
Thanks, huge improvement!

1. One part of new text isn't clear.  What is the value of `.key.name` for non-printing chars?  Exhaustive values are only listed for properties, not for .name.

The [new text](https://github.com/zauberzeug/nicegui/pull/3010/commits/fcf0a01381312f7e08c753a2f895b4aad55aaaf4) says `e.g. 'Enter'` for key.name, while key properties uses lowercase `.enter`.  Does this correlation hold for other keys?  If `evt.key.left_arrow` is true, then `evt.key.name` is ...
- LeftArrow
- Left_Arrrow
- Leftarrow
- something else?

2. Same question for key.code.  Only a few examples are given.  How do we find .code values for other keys?

3. Also, the docs use checks like `evt.key == 'f'` not `evt.key.name == 'f'`.  Presumably the implementation of == op uses key.name.  Is that correct?  If so , might be worth explicitly stating.

**Comment by ed2050** [comment]
# Example

Here's an example of why having key.name values for all keys matters.  Checking key.name is more flexible than checking properties.  For instance, I can define a dict mapping keys to actions, and use that in keyboard listener to dispatch key events.  Something like:
```py
keymappings = {
        'a' : lambda : keepnames (allnames) ,
        'o' : lambda : keepnames (allnames [0]) ,
        'p' : lambda : prune () ,
        'z' : lambda : reload_page () ,
        'LeftArrow'  : load_prev ,
        'RightArrow' : load_next ,
}

def keylisten (evt) :
    func = keymappings.get (evt.key.name, none)
    if func :  func ()

ui.keyboard (on_key = keylisten)
```
This is a convenient and dynamic way to dispatch key actions without writing `if key == 'foo' ; elif key == 'bar' ; elif ...` tests.   It breaks if certain keys like arrows don't have a unique `key.name` string and require checking `key.left_arrow` instead.

**Comment by falkoschindler** [comment]
Hi @ed2050,

The implementation of keyboard event arguments is pretty straightforward and can be found here:
https://github.com/zauberzeug/nicegui/blob/5a25f18cc2122555ac20e547a1b2e4f366d7165b/nicegui/events.py#L137-L353

It is derived from JavaScript's keyboard events which are documented here:
https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent

Maybe you can take a look and let us know what specific information should be added to the documentation? As usual, we like to keep it short, but avoiding confusion by giving a few more hints can't hurt. Thanks! 🙂

**Reply by ed2050** [reply]
Sure no problem.  I suspected they came from JS.  In that case, no need to list them all in your docs, a link will suffice.  Not the page you linked but the [page of key values](https://developer.mozilla.org/en-US/docs/Web/API/UI_Events/Keyboard_event_key_values).

Line 40 of nicegui/elements/keyboard.py could be changed to this:
>  - `name`: the name of the key (e.g. 'a', 'Enter').  An exhaustive list of possible key names can be found at https://developer.mozilla.org/en-US/docs/Web/API/UI_Events/Keyboard_event_key_values.

There are many more keys on the reference page than nicegui implements.  Keys like Function (Fn), media keys (Play, Pause, Stop), controls (Power, Brightness, Mute, Volume), etc.  If you don't plan to support those keys, add a note which keys nicegui does support, like "101-key ANSI" or whatever (I don't know the technical names).

**Reply by falkoschindler** [reply]
Ok, I added a link to that page in 53af516a403b2109ef2129b5e0c6539414d3b8b6. But I struggle to find a description for the set of keys covered with properties (`backspace`, `tab`, `enter`, ...). Sure, the list is not complete, but I find it hard to draw a line. Many of them mentioned in the [JavaScript documentation](https://developer.mozilla.org/en-US/docs/Web/API/UI_Events/Keyboard_event_key_values) have footnotes and limitations, and would obscure the autocompletion for "useful" ones. And even if a key doesn't have its own Python property, the event still works and you can simply check the key name or code via string comparison.

**Reply by ed2050** [reply]
@falkoschindler sure that's fine.  I was under the impression that only keys covered in nicegui docs (shift, enter, f1, etc) were available in nicegui.  Seems I was mistaken.  If keys any key received in the browser will be forwarded to nicegui as a key event, then no need to describe which are supported.

Thanks for addressing these doc improvements.  It may seem like a small thing, but it's very helpful.  Especially for people unfamiliar with the JS world. 👍

**Comment by ed2050** [comment]
It would also be nice to have a shortcut to detect if any modifier keys were pressed.  Detecting no mod keys is a common pattern.  So instead of tedious code like this :
```py
if key == 'a' and not (key.modifiers.alt or key.modifiers.ctrl or key.modifiers.meta or key.modifiers.shift) :
    do something ...
```
We could have code like this :
```py
if key == 'a' and not any (key.modifiers) :
   do something ....
```
This would just require that key.modifiers be iterable over its `__dict___.values`.  An even simpler approach could be :
```py
if key == 'a' and not key.modifiers :
    do something
```
which would require key.modifiers to convert to boolean.  (And either works with my example key dispatcher above.  `key == 'a'` is just for simplicity here.)

It's true that users could write their own function `hasmods (keyevent)` to test whether a key has any modifiers set.  However this is such a common use case that it should be considered for inclusion in nicegui.

Thanks.

**Reply by falkoschindler** [reply]
Ok, in 771553018ef9476236d56bbaee9f81f01c537f0f I added `__iter__` and `__len__` to `KeyboardModifiers` so that you can write
```py
ui.keyboard(lambda e: print(any(e.modifiers), len(e.modifiers), 'MOD' if e.modifiers else ''))
```
```
False 0 
True 1 MOD
True 2 MOD
True 3 MOD
True 4 MOD
```
